src/ts/Document.ts

```diff
@@ -0,0 +1,223 @@
+import { assignIn, cloneDeep, get as getIn, set as setIn } from 'lodash';
+import { ForeignKeyDbTransformer } from '#shared/classes';
+import { TransformDataError } from '#shared/errors';
+import type { GenericObject, MvRecord } from '#shared/types';
+import type Schema from './Schema';
+
+const DEFAULT_PROPERTY_DESCRIPTOR: PropertyDescriptor = {
+	configurable: false,
+	enumerable: false,
+	writable: false,
+};
+
+export interface DocumentConstructorOptionsData {
+	data: GenericObject;
+	isSubdocument?: boolean;
+}
+export interface DocumentConstructorOptionsRecord {
+	record: MvRecord;
+	isSubdocument?: boolean;
+}
+export interface DocumentConstructorOptionsRaw {
+	record: MvRecord;
+}
+export type DocumentConstructorOptions =
+	| DocumentConstructorOptionsData
+	| DocumentConstructorOptionsRecord
+	| DocumentConstructorOptionsRaw;
+
+export interface BuildForeignKeyDefinitionsResult {
+	filename: string;
+	entityName: string;
+	entityIds: string[];
+}
+
+/** A document object */
+class Document {
+	[key: string]: unknown;
+
+	public _id?: string;
+
+	public _raw?: MvRecord;
+
+	/** Array of any errors which occurred during transformation from the database */
+	public transformationErrors: TransformDataError[];
+
+	/** Schema instance which defined this document */
+	private readonly schema: Schema | null;
+
+	/** Record array of multivalue data */
+	private record: MvRecord;
+
+	/** Indicates whether this document is a subdocument of a composing parent */
+	private readonly isSubdocument: boolean;
+
+	public constructor(schema: Schema, options: DocumentConstructorOptions);
+	public constructor(schema: null, options: DocumentConstructorOptionsRaw);
+	public constructor(schema: Schema | null, options: DocumentConstructorOptions) {
+		const { data, isSubdocument, record } = this.parseConstructorOptions(options);
+
+		this.schema = schema;
+		this.record = [];
+		this.isSubdocument = isSubdocument;
+		this.transformationErrors = [];
+
+		Object.defineProperties(this, {
+			schema: DEFAULT_PROPERTY_DESCRIPTOR,
+			record: {
+				...DEFAULT_PROPERTY_DESCRIPTOR,
+				writable: true,
+			},
+			isSubdocument: DEFAULT_PROPERTY_DESCRIPTOR,
+			transformationErrors: DEFAULT_PROPERTY_DESCRIPTOR,
+		});
+
+		if (record != null) {
+			this.transformRecordToDocument(record);
+		}
+		// load the data passed to constructor into document instance
+		assignIn(this, data);
+	}
+
+	/** Transform document structure to multivalue array structure */
+	public transformDocumentToRecord(): MvRecord {
+		return this.schema === null
+			? getIn(this, '_raw', [])
+			: Object.entries(this.schema.paths).reduce(
+					(record, [keyPath, schemaType]) => {
+						const value = getIn(this, keyPath, null);
+						return schemaType.set(record, schemaType.cast(value));
```

Comment:
I added a call to `schemaType.cast` here of the value.  The previous behavior relied on `validate` having been called before `transformDocumentToRecord` which is not a good assumption.  Adding the `cast` here allows it to be called in any order.